doc: edit schema assert messages for clarity and grammar

dbkinder opened the pull request:
Use DX names (acrn:title) instead of element names in messages.
Add missing message in rdt_support.xsd (all asserts need an annotation
with documentation for the error message that will be reported).

Tracked-On: #7685

Signed-off-by: David B. Kinder <david.b.kinder@intel.com>

acrnsi commented:
Can one of the admins verify this patch?

amyreye commented on misc/config_tools/schema/checks/vm_types.xsd:
```diff
@@ -29,13 +29,13 @@
   <xs:assert test="count(vm[load_order = 'PRE_LAUNCHED_VM' and vm_type = 'RTVM']) &lt;= 1">
     <xs:annotation acrn:severity="warning" acrn:report-on="//vm[load_order = 'PRE_LAUNCHED_VM']/vm_type[text() = 'RTVM']">
-      <xs:documentation>Multiple pre-launched VMs {//vm[load_order = 'PRE_LAUNCHED_VM' and vm_type = 'RTVM']/name} are defined as RT VMs, but ACRN supports at most one pre-launched RT VM in one scenario. Adjust the VM types of those VMs to STANDARD_VM to fix this.</xs:documentation>
+      <xs:documentation>ACRN supports at most one pre-launched Realtime VM. Multiple pre-launched VMs {//vm[load_order = 'PRE_LAUNCHED_VM' and vm_type = 'RTVM']/name} are defined as RT VMs. Adjust the VM type of these VMs to STANDARD_VM to fix this.</xs:documentation>
```
Use the names of options in the UI?

```suggestion
      <xs:documentation>ACRN supports at most one pre-launched Real-time VM. Multiple pre-launched VMs {//vm[load_order = 'PRE_LAUNCHED_VM' and vm_type = 'RTVM']/name} are defined as RT VMs. Adjust the VM type of these VMs to Standard to fix this.</xs:documentation>
```

amyreye commented on misc/config_tools/schema/checks/vm_types.xsd:
```diff
@@ -29,13 +29,13 @@
   <xs:assert test="count(distinct-values(vm[vm_type = 'RTVM']/load_order)) &lt;= 1">
     <xs:annotation acrn:severity="warning" acrn:report-on="//vm/vm_type[text() = 'RTVM']">
-      <xs:documentation>Pre-launched RT VM(s) {//vm[load_order = 'PRE_LAUNCHED_VM' and vm_type = 'RTVM']/name} and post-launched RT VMs {//vm[load_order = 'POST_LAUNCHED_VM' and vm_type = 'RTVM']/name} cannot coexist. Adjust the VM types of those VMs to STANDARD_VM to fix this.</xs:documentation>
+      <xs:documentation>Pre-launched RT VM(s) {//vm[load_order = 'PRE_LAUNCHED_VM' and vm_type = 'RTVM']/name} and post-launched RT VMs {//vm[load_order = 'POST_LAUNCHED_VM' and vm_type = 'RTVM']/name} may not coexist. Adjust the VM type of these VMs to STANDARD_VM to fix this.</xs:documentation>
```
Use the names of options in the UI?

```suggestion
      <xs:documentation>Pre-launched RT VM(s) {//vm[load_order = 'PRE_LAUNCHED_VM' and vm_type = 'RTVM']/name} and post-launched RT VMs {//vm[load_order = 'POST_LAUNCHED_VM' and vm_type = 'RTVM']/name} may not coexist. Adjust the VM type of these VMs to Standard to fix this.</xs:documentation>
```

amyreye commented on misc/config_tools/schema/checks/vm_types.xsd:
```diff
@@ -7,7 +7,7 @@
   <xs:assert test="count(/acrn-config/vm[.//load_order = 'SERVICE_VM']) &lt; 2">
     <xs:annotation acrn:severity="error" acrn:report-on="//vm[.//load_order = 'SERVICE_VM']">
-      <xs:documentation>There can be at most one service VM, but more than one was configured. Verify there is only one VM with vm_type set to SERVICE_VM. </xs:documentation>
+      <xs:documentation>There may be at most one Service VM, but more than one was configured. Verify only one VM's VM type is set to SERVICE_VM. </xs:documentation>
```
Would this error occur in the UI? If so, describe how to fix in UI?

dbkinder replied:
The VM load_order option is not visible in the UI so I suspect the only way this error would show up is if the schema file were manually edited and then the schema were loaded back into the configurator and saved again.  @junjiemao1 @NanlinXie What are the expectations on the configurator if a schema is read back in that contains an error such as this and there's no way for the developer to use the configurator to fix the problem?

junjiemao1 replied:
Generally, one can only fix a manually-corrupted scenario file by manually rolling back the change. I don't think the configurator can have a systematic way to fix those issues.

What the configurator may do is to refuse the loading of such XMLs.

amyreye commented on misc/config_tools/schema/checks/virtio_devices.xsd:
```diff
@@ -10,7 +10,7 @@
                    ">
     <xs:annotation acrn:severity="error" acrn:report-on="$vm">
-      <xs:documentation>"{$vm/name}" repeats a vsock CID assignment: {$vsock}, either with this VM or with other VMs. Remove the duplicates.</xs:documentation>
+      <xs:documentation>"{$vm/name}" repeats a vsock CID assignment: {$vsock}, used in this VM or another VM. Make all vsock CID values unique.</xs:documentation>
```
```suggestion
      <xs:documentation>VM "{$vm/name}" repeats a vsock CID assignment: {$vsock}, used in this VM or another VM. Make all vsock CID values unique.</xs:documentation>
```

amyreye commented on misc/config_tools/schema/checks/rdt_support.xsd:
```diff
@@ -71,31 +71,33 @@
-
-This error cannot be fixed by tweaking the configurations. Report to _GitHub:
-https://github.com/projectacrn/acrn-hypervisor/issues if you meet this.</xs:documentation>
+pre-launched VM, such as the guest PCI hole which resides at 2GB - 3.5GB.
```
Change all mentions of "RT VM" to "RTVM" per ACRN glossary (or spell out real-time for clarity)? https://projectacrn.github.io/latest/glossary.html#term-RTVM

amyreye commented on misc/config_tools/schema/checks/rdt_support.xsd:
```diff
@@ -27,15 +27,15 @@
                   ">
     <xs:annotation>
-      <xs:documentation>For a vCAT VM, number of clos/vcpu_clos elements must be greater than 1!</xs:documentation>
+      <xs:documentation>For a VM with Virtual Cache Allocation Technology enabled, you must define more then one clos/vcpu_clos element.</xs:documentation>
```
```suggestion
      <xs:documentation>For a VM with Virtual Cache Allocation Technology enabled, you must define more than one clos/vcpu_clos element.</xs:documentation>
```

amyreye commented on misc/config_tools/schema/checks/rdt_support.xsd:
```diff
@@ -27,15 +27,15 @@
                   ">
     <xs:annotation>
-      <xs:documentation>For a vCAT VM, number of clos/vcpu_clos elements must be greater than 1!</xs:documentation>
+      <xs:documentation>For a VM with Virtual Cache Allocation Technology enabled, you must define more then one clos/vcpu_clos element.</xs:documentation>
```
We removed clos/vcpu_clos in the UI. Would any clos-related errors occur in the UI? If so, how would a user recover? This comment applies to all messages that mention clos or vcpu_clos below.

dbkinder replied:
Alas, I have no idea how the user would recover from this assertion failure since they have no way to change these settings with the current UI.  @junjiemao1 any ideas?

junjiemao1 replied:
That assertion should be dropped as those config items involved no longer exist.

dbkinder replied:
OK.  I'll delete the assert

amyreye commented on misc/config_tools/schema/checks/rdt_support.xsd:
```diff
@@ -7,15 +7,15 @@
                    else true()">
     <xs:annotation>
-      <xs:documentation>Per VM virtual_cat_support can be set only when RDT_ENABLED is 'y' and VCAT_ENABLED is 'y'.</xs:documentation>
+      <xs:documentation>Per VM Virtual Cache Allocation Technology may be enabled only when hypervisor Intel Resource Director Technology and Virtual Cache Allocation Technology are both enabled.</xs:documentation>
```
```suggestion
      <xs:documentation>VM Virtual Cache Allocation Technology may be enabled only when hypervisor Intel Resource Director Technology and Virtual Cache Allocation Technology are both enabled.</xs:documentation>
```

amyreye commented on misc/config_tools/schema/checks/rdt_support.xsd:
```diff
@@ -7,15 +7,15 @@
                    else true()">
     <xs:annotation>
-      <xs:documentation>vCAT can be enabled only when RDT_ENABLED is 'y' and CDP_ENABLED is 'n'</xs:documentation>
+      <xs:documentation>Hypervisor Virtual Cache Allocation Technology may be enabled only when hypervisor Intel Resource Director Technology is enabled and hypervisor Code and Data Prioritization is disabled.</xs:documentation>
```
In the UI, the checkbox is named Cache Allocation Technology. Intel Resource Director Technology is just a heading. But it is being discussed in JIRA 9058. This comment applies to all messages that mention enabling RDT below.

junjiemao1 commented on misc/config_tools/schema/checks/ivsh_memory.xsd:
```diff
@@ -10,7 +10,7 @@
                   count(/acrn-config//IVSHMEM_REGION[NAME=$REGION_NAME]/IVSHMEM_VMS/IVSHMEM_VM[VM_NAME=$VM_NAME]) = 1">
     <xs:annotation acrn:severity="error" acrn:report-on="$REGION_NAME">
-      <xs:documentation>InterVM shared memory region name "{$REGION_NAME}" should be different for the same VM "{$VM_NAME}"</xs:documentation>
+        <xs:documentation>VM "{$VM_NAME}" may not specify InterVM shared memory region name "{$REGION_NAME}" more than once.</xs:documentation>
```
The original description is quite confusing. What this assertions actually means is that "VM {$VM_NAME} cannot be duplicated in the list of VMs that have access to the inter-VM shared memory region {$REGION_NAME}"

dbkinder replied:
got it.  thanks

junjiemao1 commented on misc/config_tools/schema/checks/vuart_config.xsd:
```diff
@@ -10,14 +10,14 @@
                    count($vuart_connection/endpoint[./vm_name/text()=$vm_name]) = 1">
     <xs:annotation acrn:severity="error" acrn:report-on="$vuart_connection/endpoint">
-      <xs:documentation>VM "{$vm_name}" can't connected to itself</xs:documentation>
+      <xs:documentation>The vUART configuration of VM "{$vm_name}" may not connect to itself.</xs:documentation>
```
Should we mention the vUART connection name as well?

junjiemao1 commented on misc/config_tools/schema/checks/cpu_assignment.xsd:
```diff
@@ -27,28 +27,28 @@
                    count(/acrn-config/vm[@id != $pcpu/ancestor::vm//companion_vmid ]//cpu_affinity[.//pcpu_id = $pcpu]) &lt;= 1">
     <xs:annotation acrn:severity="error" acrn:report-on="//vm//cpu_affinity[.//pcpu_id = $pcpu]">
-      <xs:documentation>Physical CPU {$pcpu} is assigned to Pre-launched VM [{$pcpu/ancestor::vm/name}] and thus cannot be shared among multiple VMs. Look for, and probably remove, any affinity assignments to {$pcpu} in this VM's settings: {//vm[cpu_affinity//pcpu_id = $pcpu]/name}.</xs:documentation>
+      <xs:documentation>Physical CPU {$pcpu} is assigned to Pre-launched VM "[{$pcpu/ancestor::vm/name}]" and thus may not be shared among multiple VMs. Look for, and probably remove, any affinity assignments to CPU {$pcpu} in this VM's settings: {//vm[cpu_affinity//pcpu_id = $pcpu]/name}.</xs:documentation>
```
Some recent discussions on requirements reveal that this check is outdated and should be dropped (by a PR to be submitted soon). Please skip your edits in this PR to avoid conflicts. Thanks.

dbkinder commented:
@NanlinXie I've rebased to catch up with other release_3.0 updates, and addressed some comments from Junjie and Amy (including deleting an obsolete check recommended by Junjie).
